apollo-relay/src/lib.rs

```diff
@@ -0,0 +1,532 @@
+        ///
+        /// The relay will attempt to make the transfer 5 times before failing. If
+        /// the relay fails, the data is discarded.
+        pub fn new(addr: SocketAddr) -> Self {
+            // Spawn a task which will check if there are reports to
+            // submit every interval.
+            let tpq: Arc<Mutex<HashMap<ReporterGraph, HashMap<String, report::TracesAndStats>>>> =
+                Arc::new(Mutex::new(HashMap::new()));
+            let task_tpq = tpq.clone();
+            let (tx, mut rx) = tokio::sync::mpsc::channel::<()>(100);
+            tokio::task::spawn(async move {
+                let client = Client::new();
+                let mut interval = interval(Duration::from_secs(5));
+                interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+                interval.tick().await;
+                loop {
+                    tokio::select! {
+                        biased;
+                        mopt = rx.recv() => {
+                            match mopt {
+                                Some(_msg) => {
+                                    relay_tpq(&client, task_tpq.clone()).await;
+                                },
+                                None => break
+                            }
+                        },
+                        _ = interval.tick() => {
+                            relay_tpq(&client, task_tpq.clone()).await;
+                        }
+                    };
+                }
+            });
+            Self { addr, tpq, tx }
+        }
+
+        /// Start serving requests.
+        pub async fn serve(self) -> Result<(), Error> {
+            let addr = self.addr;
+            Server::builder()
+                .add_service(ReporterServer::new(self))
+                .serve(addr)
+                .await
+        }
+
+        async fn submit_report(
+            client: &Client,
+            key: String,
+            report: report::Report,
+        ) -> Result<Response<ReporterResponse>, Status> {
+            tracing::debug!("submitting report: {:?}", report);
+            // Protobuf encode message
+            let mut content = BytesMut::new();
+            report
+                .encode(&mut content)
+                .map_err(|e| Status::invalid_argument(e.to_string()))?;
+            // Create a gzip encoder
+            let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
+            // Write our content to our encoder
+            encoder
+                .write_all(&content)
+                .map_err(|e| Status::internal(e.to_string()))?;
+            // Finish encoding and retrieve content
+            let compressed_content = encoder
+                .finish()
+                .map_err(|e| Status::internal(e.to_string()))?;
+            let mut backoff = 0;
+            for i in 0..4 {
+                let res = client
+                    .post("https://usage-reporting.api.apollographql.com/api/ingress/traces")
```

Comment:
You will want some way to override this URL (not necessarily publicly documented) for dogfooding in our staging environment.

Reply:
Yes. Perhaps a "hidden" start-up flag?

Reply:
I've added an undocumented environment variable: APOLLO_INGRESS. You can use it as:
APOLLO_INGRESS=http://localhost:8080/api/ingress/traces router <blah>